KTOR-6497 Add documentation for opentelemetry-ktor

vnikolova opened the pull request:
[KTOR-6497](https://youtrack.jetbrains.com/issue/KTOR-6497) 

Add 2 new topics to cover the client and server third-party plugins.

marychatte commented on codeSnippets/snippets/opentelemetry/server/src/main/kotlin/Monitoring.kt:
```diff
@@ -0,0 +1,46 @@
+            call.respondText("Hello World!")
+        }
+        
+        post("/post") {
```
I think we need to write server routes: `/known-methods`, `/captured-headers` etc

https://github.com/ktorio/ktor-samples/blob/main/opentelemetry/server/src/main/kotlin/opentelemetry/ktor/example/plugins/routing/Routing.kt

marychatte commented on topics/client-opentelemetry.md:
```diff
@@ -0,0 +1,77 @@
+captured, and how span kinds are determined.
+
+> For more information on these concepts, see the
+> [OpenTelemetry tracing documentation](https://opentelemetry.io/docs/concepts/signals/traces/).
```
It might be a good idea to mention the available functions here and show how they work: attributesExtractor, knownMethods, etc. Because OpenTelemetry tracing documentation is common, and not only for Ktor

vnikolova replied:
Is there somewhere I can find the list of available functions or can you tell me which ones I've missed?

marychatte replied:
Oh, sorry. I can the links like `<include from="server-opentelemetry.md" element-id="config-known-methods"/>` so everything is mentioned

marychatte reviewed (commented):
Great work 🎉

I wrote several minor notes, but it's a good doc.
I have a suggestion to add an example with Grafana, because we won't need to turn off metrics in this case. 
@osipxd told me about [grafana-otel-lgtm docker image](https://hub.docker.com/r/grafana/otel-lgtm)

vnikolova commented:
Thanks for the suggestions @marychatte. I've now replaced Jaeger with Grafana, updated the sample project and included the section on programmatic configuration. Can you check everything one more time? 🙏

marychatte commented on topics/client-opentelemetry.md:
```diff
@@ -0,0 +1,77 @@
+Once you have %plugin_name% installed and configured, you can verify that spans are being created and propagated by
+sending requests to a service that also has telemetry enabled—such as one using
+[`KtorServerTelemetry`](server-opentelemetry.md). Viewing both sides of the trace in an observability backend like
+Jaeger, Zipkin, or Grafana Tempo will confirm that distributed tracing is working end-to-end.
```
minor: maybe add links to Jaeger, Zipkin, or Grafana Tempo here?
(i'm a really big fan of links 😅)

vnikolova replied:
That's good because I keep missing them 😄

marychatte reviewed (approved):
Great work! I like it :)

zamulla commented on topics/server-opentelemetry.md:
```diff
@@ -0,0 +1,327 @@
+## Install %plugin_name% {id="install_plugin"}
+
+To [install](server-plugins.md#install) the `%plugin_name%` plugin to the application, pass it to the `install` function
+in the specified [module](server-modules.md) and set the [configured `OpenTelemetry` instance](#configure-otel):
```
Are the links to "install" and "module" helpful here? If there is context that may be necessary, I would provide it here, but I also don't think that there is.

zamulla replied:
If it's the same everywhere, then yeah, that's a separate thing. But I do think there's little reason to follow these links from this particular page, even if I understand what's on the other side.

zamulla commented on topics/server-opentelemetry.md:
```diff
@@ -0,0 +1,327 @@
+
+The `%plugin_name%` plugin enables distributed tracing of incoming HTTP requests in a Ktor server application. It
+automatically creates spans containing route, HTTP method, and status code information, extracts existing trace context
+from incoming request headers, and allows customization of span names, attributes, and span kinds.
```
Should we link the definition of "span"? Maybe not, if our primary audience is OpenTelemetry users looking for a server, but also maybe yes, if our audience is Ktor users looking for a telemetry framework.

zamulla commented on topics/server-opentelemetry.md:
```diff
@@ -0,0 +1,327 @@
+
+### Add custom attributes {id="config-custom-attributes"}
+
+To attach custom attributes at the start and end of a span, use the `attributesExtractor` property:
```
```suggestion
To attach custom attributes at the start or end of a span, use the `attributesExtractor` property:
```

Should this be "or"?

zamulla commented on topics/server-opentelemetry.md:
```diff
@@ -0,0 +1,327 @@
+[Tempo](https://grafana.com/oss/tempo/) (traces), [Loki](https://grafana.com/oss/loki/) (logs), and
+[Mimir](https://grafana.com/oss/mimir/) (metrics).
+
+### Using Docker Compose
```
Should this and the CLI section be tabs instead? If they show different ways of doing exactly the same thing.

vnikolova replied:
Or perhaps the Docker CLI chapter can be collabsible? 🤔 We have a similar case in [this tutorial](https://ktor.io/docs/server-create-a-new-project.html#create_project_with_ktor_cli_tool).

zamulla replied:
I think with the new project instructions there are a lot of screenshots to be collapsed and here a collapsed section can be more annoying than convenient — although I have no idea whether Docker CLI is a niche use case or not.

zamulla commented on topics/server-opentelemetry.md:
```diff
@@ -0,0 +1,327 @@
+
+### Accessing Grafana UI
+
+Once running, the Grafana UI will be available at [http://127.0.0.1:3000/](http://127.0.0.1:3000/).
```
In the readme file for the opentelemetry sample project, the address is `localhost:3000`. I don't know which one is preferable, but they should probably be the same :)

vnikolova replied:
Good catch! The Grafana docs state `http://localhost:3000/`  as default so I'll use that.

zamulla reviewed (commented):
Looks good! I have some suggestions on phrasing, but mostly questions — let me know what you think.

vnikolova commented on topics/server-opentelemetry.md:
```diff
@@ -0,0 +1,327 @@
+This approach allows you to define exporters, processors, and propagators in code, instead of relying on
+environment-based configuration.
+
+The following example shows how to configure OpenTelemetry manually with an OTLP exporter, a span processor, and a trace
```
OpenTelemetry calls them "zero-code autoconfigure" and "programmatic": https://opentelemetry.io/docs/languages/java/configuration/#programmatic-configuration 
So I think it's better to replace "manually" with "programmatically"

vnikolova commented on topics/server-opentelemetry.md:
```diff
@@ -0,0 +1,327 @@
+automatically creates spans containing route, HTTP method, and status code information, extracts existing trace context
+from incoming request headers, and allows customization of span names, attributes, and span kinds.
+
+> On the client, OpenTelemetry provides the [KtorClientTelemetry](client-opentelemetry.md) plugin, which instruments
```
As a newcomer to observability tools I didn't know it either, but it seems to be the correct term. I will reword the note though.
